README.md
Lita::Handlers::Random
======================

[![Gem Version](https://badge.fury.io/rb/lita-random.svg)](http://badge.fury.io/rb/lita-random)
[![Build Status](https://travis-ci.org/braiden-vasco/lita-random.svg)](https://travis-ci.org/braiden-vasco/lita-random)
[![Coverage Status](https://coveralls.io/repos/braiden-vasco/lita-random/badge.svg)](https://coveralls.io/r/braiden-vasco/lita-random)

Generator of random numbers and strings for the [Lita](http://lita.io) chat bot.

Usage
-----

At first, see the documentation for Lita: http://docs.lita.io/

### Installation

Add **lita-random** to your Lita instance's Gemfile:

```ruby
gem 'lita-random', '~> 0.6.0'
```

### Commands

The space between instructions can be skipped, so you can write
`random hex` as `randomhex`, `rand b64` as `randb64`

Arguments for all commands (excluding `random case`, `shuffle` and `sample`)
can be passed as keyword arguments too. Put them after
positional arguments with two dashes before, or the first letter
of argument name with one dash. Examples:

```
You: lita random 5 10
Lita: 7
You: lita random 5 --to 5
Lita: 9
You: lita random 5 -t 10
Lita: 5
You: lita random --to 10 --from 5
Lita: 6
You: lita random -f 5 -t 10
Lita: 9
```

Command `random` has alias `rand`

* `random` -
  random float number, greater or equal to 0 and lesser than 1

* `random <to>` -
  random integer or float number, greater or equal to 0 and lesser than `to`

* `random <from> <to>` -
  random integer or float number, greater or equal to `from` and lesser than `to`

* `random case <string>` -
  randomize case of each character of string `string`

Command `base64` has alias `b64`

* `random base64 <size=16>` -
  random base64 string, length of source string is `size`,
  length of result is about `size * 4 / 3` (24 with default value of `size`)

Command `hex` has aliases `hx` and `x`

* `random hex <size=16>` -
  random hexadecimal string with length `size * 2`

Command `uuid` has alias `uid`

* `random uuid` -
  v4 random UUID (Universally Unique IDentifier). The version 4 UUID
  is purely random (except the version). It doesn’t contain
  meaningful information such as MAC address, time, etc.

Command `password` has alias `pass`

* `random password <length=16>` -
  random password with length `length` containing characters
  in upper and lower case, and digits

* `random smart password <length=8>` -
  random pronounceable password with a minimum length of `length`

* `shuffle <array, ...>` -
  new array with elements of `array` shuffled

* `sample <count=1> <array, ...>` -
  choose `count` random elements from `array`

### Examples

```
You: lita random
Lita: 0.52693215290866
You: lita rand
Lita: 0.8332672468501509

You: lita random 5
Lita: 2
You: lita rand --to 100
Lita: 24

You: lita random -f 1.5
Lita: 1.1828046952104034
You: lita random 10 --from 5
Lita: 8
You: lita rand 1000 2000
Lita: 1240
You: lita random 12 13.5
Lita: 13.458799783677263

You: lita random case foo bar car cdr
Lita: FoO bAr CAr Cdr

You: lita random base64
Lite: 56wnZVVVStKG5+I6L4LVGw==
You: lita randb64 10
Lita: NMkIhB+iG2VNpg==

You: lita random hex
Lita: 7d6f39cf53303db862bac5e1b36eb0fa
You: lita randx -s 10
Lita: 0d8297837d6e63b8c25b

You: lita random uuid
Lita: f1281565-3c8a-475d-a1aa-565ff840a42c

You: lita random password
Lita: TXXBu7XEBdwFF7fP
You: lita randpass --length 8
Lita: 7lFateFs

You: lita random smart password
Lita: mospowot
You: lita random smartpass 16
Lita: jophemyrdivezagi

You: lita shuffle 1, 2, 3, 4, 5
Lita: 3, 1, 5, 2, 4

You: lita sample 1, 2, 3, 4, 5
Lita: 4
You: lita sample 3  1, 2, 3, 4, 5
Lita: 5, 1, 2
```


CHANGELOG.md
# Change Log

## [v0.6.0](https://github.com/braiden-vasco/lita-random/tree/v0.6.0) (2015-10-24)
[Full Changelog](https://github.com/braiden-vasco/lita-random/compare/v0.5.1...v0.6.0)

**Merged pull requests:**

- Support keyword arguments partially [\#46](https://github.com/braiden-vasco/lita-random/pull/46) ([braiden-vasco](https://github.com/braiden-vasco))
- Test like Lita [\#45](https://github.com/braiden-vasco/lita-random/pull/45) ([braiden-vasco](https://github.com/braiden-vasco))
- Test with Rubinius [\#44](https://github.com/braiden-vasco/lita-random/pull/44) ([braiden-vasco](https://github.com/braiden-vasco))
- Don't use version label and version PRs [\#43](https://github.com/braiden-vasco/lita-random/pull/43) ([braiden-vasco](https://github.com/braiden-vasco))

## [v0.5.1](https://github.com/braiden-vasco/lita-random/tree/v0.5.1) (2015-06-25)
[Full Changelog](https://github.com/braiden-vasco/lita-random/compare/v0.5.0...v0.5.1)

**Merged pull requests:**

- Test with multiple versions of Ruby [\#39](https://github.com/braiden-vasco/lita-random/pull/39) ([braiden-vasco](https://github.com/braiden-vasco))
- Remove open-ended dependencies [\#38](https://github.com/braiden-vasco/lita-random/pull/38) ([braiden-vasco](https://github.com/braiden-vasco))
- Remove unnecessary dependencies [\#37](https://github.com/braiden-vasco/lita-random/pull/37) ([braiden-vasco](https://github.com/braiden-vasco))

## [v0.5.0](https://github.com/braiden-vasco/lita-random/tree/v0.5.0) (2015-06-24)
[Full Changelog](https://github.com/braiden-vasco/lita-random/compare/v0.4.0...v0.5.0)

**Merged pull requests:**

- Add command "random case" [\#34](https://github.com/braiden-vasco/lita-random/pull/34) ([braiden-vasco](https://github.com/braiden-vasco))
- Add command "sample" [\#33](https://github.com/braiden-vasco/lita-random/pull/33) ([braiden-vasco](https://github.com/braiden-vasco))
- Add command "shuffle" [\#31](https://github.com/braiden-vasco/lita-random/pull/31) ([braiden-vasco](https://github.com/braiden-vasco))
- Move development dependencies to gemspec [\#30](https://github.com/braiden-vasco/lita-random/pull/30) ([braiden-vasco](https://github.com/braiden-vasco))

## [v0.4.0](https://github.com/braiden-vasco/lita-random/tree/v0.4.0) (2015-06-23)
[Full Changelog](https://github.com/braiden-vasco/lita-random/compare/v0.3.0...v0.4.0)

**Merged pull requests:**

- Add password generator [\#28](https://github.com/braiden-vasco/lita-random/pull/28) ([braiden-vasco](https://github.com/braiden-vasco))
- Add smart password generator [\#27](https://github.com/braiden-vasco/lita-random/pull/27) ([braiden-vasco](https://github.com/braiden-vasco))
- Add help for commands [\#26](https://github.com/braiden-vasco/lita-random/pull/26) ([braiden-vasco](https://github.com/braiden-vasco))

## [v0.3.0](https://github.com/braiden-vasco/lita-random/tree/v0.3.0) (2015-06-21)
[Full Changelog](https://github.com/braiden-vasco/lita-random/compare/v0.2.0...v0.3.0)

**Merged pull requests:**

- Add UUID generator [\#23](https://github.com/braiden-vasco/lita-random/pull/23) ([braiden-vasco](https://github.com/braiden-vasco))
- Optional spaces between commands, add more aliases [\#22](https://github.com/braiden-vasco/lita-random/pull/22) ([braiden-vasco](https://github.com/braiden-vasco))
- Add hexadecimal string generator [\#21](https://github.com/braiden-vasco/lita-random/pull/21) ([braiden-vasco](https://github.com/braiden-vasco))
- Add base64 generator [\#20](https://github.com/braiden-vasco/lita-random/pull/20) ([braiden-vasco](https://github.com/braiden-vasco))
- Refactor regular expressions [\#19](https://github.com/braiden-vasco/lita-random/pull/19) ([braiden-vasco](https://github.com/braiden-vasco))

## [v0.2.0](https://github.com/braiden-vasco/lita-random/tree/v0.2.0) (2015-06-21)
[Full Changelog](https://github.com/braiden-vasco/lita-random/compare/v0.1.0...v0.2.0)

**Merged pull requests:**

- Allow float ranges [\#17](https://github.com/braiden-vasco/lita-random/pull/17) ([braiden-vasco](https://github.com/braiden-vasco))
- Basic testing [\#15](https://github.com/braiden-vasco/lita-random/pull/15) ([braiden-vasco](https://github.com/braiden-vasco))

## [v0.1.0](https://github.com/braiden-vasco/lita-random/tree/v0.1.0) (2015-06-20)
**Merged pull requests:**

- Add instructions [\#11](https://github.com/braiden-vasco/lita-random/pull/11) ([braiden-vasco](https://github.com/braiden-vasco))
- Basic implementation [\#10](https://github.com/braiden-vasco/lita-random/pull/10) ([braiden-vasco](https://github.com/braiden-vasco))
- Initialize Lita handler [\#9](https://github.com/braiden-vasco/lita-random/pull/9) ([braiden-vasco](https://github.com/braiden-vasco))
- Add changelog generator [\#8](https://github.com/braiden-vasco/lita-random/pull/8) ([braiden-vasco](https://github.com/braiden-vasco))
- Use YARD [\#7](https://github.com/braiden-vasco/lita-random/pull/7) ([braiden-vasco](https://github.com/braiden-vasco))
- Configure Coveralls [\#6](https://github.com/braiden-vasco/lita-random/pull/6) ([braiden-vasco](https://github.com/braiden-vasco))
- Configure Travis CI [\#5](https://github.com/braiden-vasco/lita-random/pull/5) ([braiden-vasco](https://github.com/braiden-vasco))
- Add README [\#4](https://github.com/braiden-vasco/lita-random/pull/4) ([braiden-vasco](https://github.com/braiden-vasco))
- Configure testing [\#3](https://github.com/braiden-vasco/lita-random/pull/3) ([braiden-vasco](https://github.com/braiden-vasco))
- Configure linting [\#2](https://github.com/braiden-vasco/lita-random/pull/2) ([braiden-vasco](https://github.com/braiden-vasco))
- Initialize gem [\#1](https://github.com/braiden-vasco/lita-random/pull/1) ([braiden-vasco](https://github.com/braiden-vasco))



\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

spec/spec_helper.rb
# This should be on the top of the file
require 'simplecov'

# rubocop:disable Style/BlockComments

require 'lita-random'
require 'lita/rspec'

# A compatibility mode is provided for older plugins upgrading from Lita 3.
# Since this plugin was generated with Lita 4, the compatibility mode
# should be left disabled.
Lita.version_3_compatibility_mode = false

# This file was generated by the `rspec --init` command. Conventionally, all
# specs live under a `spec` directory, which RSpec adds to the `$LOAD_PATH`.
# The generated `.rspec` file contains `--require spec_helper` which will cause
# this file to always be loaded, without a need to explicitly require it in any
# files.
#
# Given that it is always loaded, you are encouraged to keep this file as
# light-weight as possible. Requiring heavyweight dependencies from this file
# will add to the boot time of your test suite on EVERY test run, even for an
# individual file that may not need all of that loaded. Instead, consider making
# a separate helper file that requires the additional dependencies and performs
# the additional setup, and require it from the spec files that actually need
# it.
#
# The `.rspec` file also contains a few flags that are not defaults but that
# users commonly want.
#
# See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration
RSpec.configure do |config|
  config.before do
    registry.register_hook(:trigger_route, Lita::Extensions::KeywordArguments)
  end

  # rspec-expectations config goes here. You can use an alternate
  # assertion/expectation library such as wrong or the stdlib/minitest
  # assertions if you prefer.
  config.expect_with :rspec do |expectations|
    # This option will default to `true` in RSpec 4. It makes the `description`
    # and `failure_message` of custom matchers include text for helper methods
    # defined using `chain`, e.g.:
    #     be_bigger_than(2).and_smaller_than(4).description
    #     # => "be bigger than 2 and smaller than 4"
    # ...rather than:
    #     # => "be bigger than 2"
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  # rspec-mocks config goes here. You can use an alternate test double
  # library (such as bogus or mocha) by changing the `mock_with` option here.
  config.mock_with :rspec do |mocks|
    # Prevents you from mocking or stubbing a method that does not exist on
    # a real object. This is generally recommended, and will default to
    # `true` in RSpec 4.
    mocks.verify_partial_doubles = true
  end

# The settings below are suggested to provide a good initial experience
# with RSpec, but feel free to customize to your heart's content.
=begin
  # These two settings work together to allow you to limit a spec run
  # to individual examples or groups you care about by tagging them with
  # `:focus` metadata. When nothing is tagged with `:focus`, all examples
  # get run.
  config.filter_run :focus
  config.run_all_when_everything_filtered = true

  # Allows RSpec to persist some state between runs in order to support
  # the `--only-failures` and `--next-failure` CLI options. We recommend
  # you configure your source control system to ignore this file.
  config.example_status_persistence_file_path = "spec/examples.txt"

  # Limits the available syntax to the non-monkey patched syntax that is
  # recommended. For more details, see:
  #   - http://myronmars.to/n/dev-blog/2012/06/rspecs-new-expectation-syntax
  #   - http://www.teaisaweso.me/blog/2013/05/27/rspecs-new-message-expectation-syntax/
  #   - http://myronmars.to/n/dev-blog/2014/05/notable-changes-in-rspec-3#new__config_option_to_disable_rspeccore_monkey_patching
  config.disable_monkey_patching!

  # This setting enables warnings. It's recommended, but in some cases may
  # be too noisy due to issues in dependencies.
  config.warnings = true

  # Many RSpec users commonly either run the entire suite or an individual
  # file, and it's useful to allow more verbose output when running an
  # individual spec file.
  if config.files_to_run.one?
    # Use the documentation formatter for detailed output,
    # unless a formatter has already been configured
    # (e.g. via a command-line flag).
    config.default_formatter = 'doc'
  end

  # Print the 10 slowest examples and example groups at the
  # end of the spec run, to help surface which specs are running
  # particularly slow.
  config.profile_examples = 10

  # Run specs in random order to surface order dependencies. If you find an
  # order dependency and want to debug it, you can fix the order by providing
  # the seed, which is printed after each run.
  #     --seed 1234
  config.order = :random

  # Seed global randomization in this process using the `--seed` CLI option.
  # Setting this allows you to use `--seed` to deterministically reproduce
  # test failures related to randomization by passing the same `--seed` value
  # as the one that triggered the failure.
  Kernel.srand config.seed
=end
end


spec/lita/handlers/random_spec/random_spec.rb
describe Lita::Handlers::Random, lita_handler: true do
  it { is_expected.to route_command('random').to :route_random }
  it { is_expected.to route_command('rand').to :route_random }
  it { is_expected.to route_command('RaNdOm').to :route_random }
  it { is_expected.to route_command('rAnD').to :route_random }

  it { is_expected.to route_command('random  5').to :route_random }
  it { is_expected.to route_command('rand   17').to :route_random }
  it { is_expected.to route_command('rAnDoM  59827').to :route_random }
  it { is_expected.to route_command('RanD  0').to :route_random }

  it { is_expected.to route_command('random  5.0').to :route_random }
  it { is_expected.to route_command('rand 17.74826').to :route_random }
  it { is_expected.to route_command('rAnDoM 5927.1').to :route_random }
  it { is_expected.to route_command('RanD  0.12397').to :route_random }

  it { is_expected.to route_command('random    0  1').to :route_random }
  it { is_expected.to route_command('rand 102  2695').to :route_random }
  it { is_expected.to route_command('rANdOM  1643 0').to :route_random }
  it { is_expected.to route_command('ranD  10   100').to :route_random }

  it { is_expected.to route_command('random  0   1.0').to :route_random }
  it { is_expected.to route_command('rand 102.0  25').to :route_random }
  it { is_expected.to route_command('rANdOM   13.1 0').to :route_random }
  it { is_expected.to route_command('ranD 10   0.5').to :route_random }

  it { is_expected.to route_command('random  --to 5').to :route_random }
  it { is_expected.to route_command('rand -t  17').to :route_random }
  it { is_expected.to route_command('rAnDoM -t 59827').to :route_random }
  it { is_expected.to route_command('RanD  --to  0').to :route_random }

  it { is_expected.to route_command('random  -t 5.0').to :route_random }
  it { is_expected.to route_command('rand --to 17.74826').to :route_random }
  it { is_expected.to route_command('rAnDoM  -t 5927.1').to :route_random }
  it { is_expected.to route_command('RanD --to  0.12397').to :route_random }

  it { is_expected.to route_command('random --from  0 -to 1').to :route_random }
  it { is_expected.to route_command('rand -f 102 --to  2695').to :route_random }
  it { is_expected.to route_command('rANdOM  -f  1643 -t  0').to :route_random }
  it { is_expected.to route_command('ranD --from  10 -t 100').to :route_random }

  it { is_expected.to route_command('random --from 0 -t 1.0').to :route_random }
  it { is_expected.to route_command('rand  -f  102.0 -to 25').to :route_random }
  it { is_expected.to route_command('rANdOM -f 13.1 -t 0').to :route_random }
  it { is_expected.to route_command('ranD --from 1 --to 0.5').to :route_random }

  it do
    is_expected.not_to route_command('rand smart pass 10 --length 8')
      .to :route_random
  end

  describe '/random' do
    it 'replies in correct format' do
      send_command 'random'
      expect(replies.last).to match(/\A\d\.\d+\z/)
      value = replies.last.to_f
      expect(value).to be >= 0
      expect(value).to be < 1
    end
  end

  describe '/random <to>' do
    it 'replies in correct format' do
      send_command 'random 1'
      expect(replies.last).to eq '0'
    end

    it 'replies in correct format' do
      send_command 'random --to 1'
      expect(replies.last).to eq '0'
    end

    it 'replies in correct format' do
      send_command 'random -t 1'
      expect(replies.last).to eq '0'
    end

    it 'replies in correct format' do
      send_command 'random 20'
      expect(replies.last).to match(/\A\d+\z/)
      value = replies.last.to_i
      expect(value).to be >= 0
      expect(value).to be < 20
    end

    it 'replies in correct format' do
      send_command 'random  -t  20'
      expect(replies.last).to match(/\A\d+\z/)
      value = replies.last.to_i
      expect(value).to be >= 0
      expect(value).to be < 20
    end

    it 'doesn\'t reply if both positional and keyword arguments present' do
      send_command 'random 2 --to 10'
      expect(replies).to be_empty
    end
  end

  describe '/random <to:float>' do
    it 'replies in correct format' do
      send_command 'random 1.0'
      expect(replies.last).to match(/\A\d\.\d+\z/)
      value = replies.last.to_f
      expect(value).to be >= 0
      expect(value).to be < 1
    end

    it 'replies in correct format' do
      send_command 'random --to 1.0'
      expect(replies.last).to match(/\A\d\.\d+\z/)
      value = replies.last.to_f
      expect(value).to be >= 0
      expect(value).to be < 1
    end

    it 'replies in correct format' do
      send_command 'random -t 1.0'
      expect(replies.last).to match(/\A\d\.\d+\z/)
      value = replies.last.to_f
      expect(value).to be >= 0
      expect(value).to be < 1
    end

    it 'replies in correct format' do
      send_command 'random 20.0'
      expect(replies.last).to match(/\A\d+\.\d+\z/)
      value = replies.last.to_i
      expect(value).to be >= 0
      expect(value).to be < 20
    end

    it 'replies in correct format' do
      send_command 'random  --to  20.0'
      expect(replies.last).to match(/\A\d+\.\d+\z/)
      value = replies.last.to_i
      expect(value).to be >= 0
      expect(value).to be < 20
    end

    it 'doesn\'t reply if both positional and keyword arguments present' do
      send_command 'random 2.0 --to 10.0'
      expect(replies).to be_empty
    end
  end

  describe '/random <from> <to>' do
    it 'replies in correct format' do
      send_command 'random 0 1'
      expect(replies.last).to eq '0'
    end

    it 'replies in correct format' do
      send_command 'random 1 -f 0'
      expect(replies.last).to eq '0'
    end

    it 'replies in correct format' do
      send_command 'random 1 --from 0'
      expect(replies.last).to eq '0'
    end

    it 'replies in correct format' do
      send_command 'random --to 1 --from 0'
      expect(replies.last).to eq '0'
    end

    it 'replies in correct format' do
      send_command 'random -t 1 -f 0'
      expect(replies.last).to eq '0'
    end

    it 'replies in correct format' do
      send_command 'random 0 20'
      expect(replies.last).to match(/\A\d+\z/)
      value = replies.last.to_i
      expect(value).to be >= 0
      expect(value).to be < 20
    end

    it 'replies in correct format' do
      send_command 'random 20  --from 0'
      expect(replies.last).to match(/\A\d+\z/)
      value = replies.last.to_i
      expect(value).to be >= 0
      expect(value).to be < 20
    end

    it 'replies in correct format' do
      send_command 'random 10010 -f 10000'
      expect(replies.last).to match(/\A\d+\z/)
      value = replies.last.to_i
      expect(value).to be >= 10_000
      expect(value).to be < 10_010
    end

    it 'doesn\'t reply if both positional and keyword arguments present' do
      send_command 'random 10 --from 2 --to 10'
      expect(replies).to be_empty
    end
  end

  describe '/random <from:float> <to:float>' do
    it 'replies in correct format' do
      send_command 'random 12 13.5'
      expect(replies.last).to match(/\A\d+\.\d+\z/)
      value = replies.last.to_f
      expect(value).to be >= 12
      expect(value).to be < 13.5
    end

    it 'replies in correct format' do
      send_command 'random 16.3 17'
      expect(replies.last).to match(/\A\d+\.\d+\z/)
      value = replies.last.to_f
      expect(value).to be >= 16.3
      expect(value).to be < 17
    end

    it 'replies in correct format' do
      send_command 'random 1.8 3.4'
      expect(replies.last).to match(/\A\d+\.\d+\z/)
      value = replies.last.to_f
      expect(value).to be >= 1.8
      expect(value).to be < 3.4
    end

    it 'replies in correct format' do
      send_command 'random 13.5  -f  12'
      expect(replies.last).to match(/\A\d+\.\d+\z/)
      value = replies.last.to_f
      expect(value).to be >= 12
      expect(value).to be < 13.5
    end

    it 'replies in correct format' do
      send_command 'random 17  --from 16.3'
      expect(replies.last).to match(/\A\d+\.\d+\z/)
      value = replies.last.to_f
      expect(value).to be >= 16.3
      expect(value).to be < 17
    end

    it 'replies in correct format' do
      send_command 'random  --to  3.4 -f  1.8'
      expect(replies.last).to match(/\A\d+\.\d+\z/)
      value = replies.last.to_f
      expect(value).to be >= 1.8
      expect(value).to be < 3.4
    end

    it 'doesn\'t reply if both positional and keyword arguments present' do
      send_command 'random 10.0 --from 2.0 --to 10.0'
      expect(replies).to be_empty
    end
  end
end


lib/lita/handlers/random.rb
require 'lita-keyword-arguments'

##
# Lita module.
#
module Lita
  ##
  # Lita handlers module.
  #
  module Handlers
    ##
    # Generator of random numbers and strings for the Lita chat bot.
    #
    class Random < Handler # rubocop:disable Metrics/ClassLength
      HELP = {
        'random' =>
          'random float number, greater or equal to 0 and lesser than 1',

        'random <to>' =>
          'random integer or float number, ' \
          'greater or equal to 0 and lesser than `to`',

        'random <from> <to>' =>
          'random integer or float number, ' \
          'greater or equal to `from` and lesser than `to`',

        'random case <string>' =>
          'randomize case of each character of string `string`',

        'random base64 <size=16>' =>
          'random base64 string, length of source string is `size`, ' \
          'length of result is about `size * 4 / 3` ' \
          '(24 with default value of `size`)',

        'random hex <size=16>' =>
          'random hexadecimal string with length `size * 2`',

        'random uuid' =>
          'v4 random UUID (Universally Unique IDentifier). ' \
          'The version 4 UUID is purely random (except the version). ' \
          'It doesn’t contain meaningful information ' \
          'such as MAC address, time, etc.',

        'random password <length=16>' =>
          'random password with length `length` containing characters ' \
          'in upper and lower case, and digits',

        'random smart password <length=8>' =>
          'random pronounceable password with a minimum length of `length`',

        'shuffle <array, ...>' =>
          'new array with elements of `array` shuffled',

        'sample <count=1> <array, ...>' =>
          'choose `count` random elements from `array`',
      }

      route(
        /^rand(om)?((\s+(?<from>\d+(\.\d+)?))?\s+(?<to>\d+(\.\d+)?))?($|\s*-)/i,
        :route_random,
        command: true, help: HELP,
        kwargs: {
          from: { short: 'f' },
          to: { short: 't' },
        }
      )

      def route_random(response)
        from = extract_argument(response, 0, :from, &method(:str_to_num)) || 0
        to = extract_argument(response, 1, :to, &method(:str_to_num)) || 1.0

      rescue RuntimeError # rubocop:disable Lint/HandleExceptions
      else
        result = if from.is_a?(Float) || to.is_a?(Float)
                   # In Rubinius,
                   # ::Random.rand can not exclude non Integer end value
                   ::Random.rand(from..(to - Float::EPSILON))
                 else
                   ::Random.rand(from...to)
                 end

        response.reply(result.to_s)
      end

      route(/^rand(om)?\s*case(\s+(?<s>.*))?$/i,
            :route_random_case, command: true)
      def route_random_case(response)
        response.reply(random_case(response.matches[0][0] || ''))
      end

      route(
        /^rand(om)?\s*b(ase)?64(\s+(?<n>\d+))?($|\s)/i,
        :route_random_base64,
        command: true,
        kwargs: {
          size: { short: 's' },
        }
      )

      def route_random_base64(response)
        size = extract_argument(response, 0, :size, &:to_i) || 16

      rescue RuntimeError # rubocop:disable Lint/HandleExceptions
      else
        response.reply(SecureRandom.base64(size))
      end

      route(
        /^rand(om)?\s*(he?)?x(\s+(?<n>\d+))?($|\s)/i,
        :route_random_hex,
        command: true,
        kwargs: {
          size: { short: 's' },
        }
      )

      def route_random_hex(response)
        size = extract_argument(response, 0, :size, &:to_i) || 16

      rescue RuntimeError # rubocop:disable Lint/HandleExceptions
      else
        response.reply(SecureRandom.hex(size))
      end

      route(/^rand(om)?\s*u?uid$/i, :route_random_uuid, command: true)
      def route_random_uuid(response)
        response.reply(SecureRandom.uuid)
      end

      route(
        /^rand(om)?\s*smart\s*pass(word)?(\s+(?<n>\d+))?($|\s)/i,
        :route_random_smart_pass,
        command: true,
        kwargs: {
          length: { short: 'l' },
        }
      )

      def route_random_smart_pass(response)
        length = extract_argument(response, 0, :length, &:to_i) || 8

      rescue RuntimeError # rubocop:disable Lint/HandleExceptions
      else
        response.reply(smart_password(length))
      end

      route(
        /^rand(om)?\s*pass(word)?(\s+(?<n>\d+))?($|\s)/i,
        :route_random_pass,
        command: true,
        kwargs: {
          length: { short: 'l' },
        }
      )

      def route_random_pass(response)
        length = extract_argument(response, 0, :length, &:to_i) || 16

      rescue RuntimeError # rubocop:disable Lint/HandleExceptions
      else
        response.reply(password(length))
      end

      route(/^shuffle(\s+(?<array>.*))?$/i, :route_shuffle, command: true)
      def route_shuffle(response)
        s = response.matches[0][0]
        a = s ? s.split(',').map(&:strip) : []
        response.reply(a.shuffle.join(', '))
      end

      route(/^sample(\s+((?<n>\d+)\s+)?(?<array>.*))?$/i,
            :route_sample, command: true)
      def route_sample(response)
        matches = response.matches[0]
        n = matches[0]
        s = matches[1]
        a = s ? s.split(',').map(&:strip) : []
        result = n ? a.sample(n.to_i).join(', ') : a.sample.to_s
        response.reply(result)
      end

    protected

      def random_case(s)
        s.each_char.map do |c|
          ::Random.rand(2).zero? ? c.upcase : c.downcase
        end.join
      end

      SMART_PASS_SEQS = {
        false => %w(
          b c d f g h j k l m n p qu r s t v w x z
          ch cr fr nd ng nk nt ph pr rd sh sl sp st th tr lt
        ),
        true => %w(a e i o u y),
      }

      def smart_password(min_length)
        password = ''
        sequence_id = false
        while password.length < min_length
          sequence = SMART_PASS_SEQS[sequence_id]
          password << sequence.sample
          sequence_id = !sequence_id
        end
        password
      end

      PASS_CHARS = [*'a'..'z', *'A'..'Z', *'0'..'9']

      def password(length)
        (0...length).map { |_| PASS_CHARS.sample }.join
      end

      def extract_argument(response, index, name)
        matches = response.matches[0]
        kwargs = response.extensions[:kwargs]

        positional = matches[index]
        kw = kwargs[name]

        fail if positional && kw

        s = positional || kw

        block_given? && s ? yield(s) : s
      end

      def str_to_num(s)
        s =~ /\./ ? s.to_f : s.to_i
      end
    end

    Lita.register_handler(Random)
  end
end
